### DISCUSSION TITLE
Can't get the custom functions to work

### DISCUSSION THREAD

**Discussion by vhpoet** [handsontable/hyperformula#980] [API Questions]
Followed your [readme](https://handsontable.github.io/hyperformula/guide/custom-functions.html#implementing-your-custom-function) to come up with the below code which throws a `Function name METRIC not recognized.`. What am I doing wrong? Here's the [codesandbox](https://codesandbox.io/s/wonderful-browser-yzzbpp?file=/src/App.js). Might be the same bug as [this](https://github.com/handsontable/hyperformula/issues/957).

```js
import HyperFormula, { FunctionPlugin } from "hyperformula";

class MetricPlugin extends FunctionPlugin {
  static implementedFunctions = {
    METRIC: {
      method: "metric"
    }
  };

  metric(ast, state) {
    return this.runFunction(ast, state, this.metadata('METRIC'), () => 63);
  }
}

export default function App() {
  HyperFormula.registerFunctionPlugin(MetricPlugin);

  const hfInstance = HyperFormula.buildFromArray([["=METRIC()"]], {
    licenseKey: "gpl-v3"
  });

  const value = hfInstance.getCellValue({ sheet: 0, col: 0, row: 0 });

  console.log(value);

  return null;
}
```

**Comment by sequba** [comment]
@vhpoet I am sorry you experienced an issue with HyperFormula. Our guide on custom functions is outdated and contains misleading information. We will fix it soon (https://github.com/handsontable/hyperformula/issues/968).

[Here](https://codesandbox.io/s/github/handsontable/hyperformula-demos/tree/2.0.x/custom-functions?file=/src/countHFFunctionPlugin.js) is a working example of a simple custom function that you can use. It does not use the helper method `runFunction` but for functions with no arguments it is not necessary.

In case of your code - it requires a few things to make it work:
- you need to add translations for your function
- if you want to use `runFunction` you must pass `ast.args` as its first parameter
- if you want to use `runFunction` you must specify the `parameters` array in the function metadata

After applying these fixes:
```js
import HyperFormula, { FunctionPlugin } from "hyperformula";

class MetricPlugin extends FunctionPlugin {
  static implementedFunctions = {
    METRIC: {
      method: "metric",
      parameters: []
    }
  };

  static translations = {
    'enGB': {
      'METRIC': 'METRIC'
    },
  }

  metric(ast, state) {
    return this.runFunction(ast.args, state, this.metadata('METRIC'), () => 63);
  }
}

export default function App() {
  HyperFormula.registerFunctionPlugin(MetricPlugin, MetricPlugin.translations);

  const hfInstance = HyperFormula.buildFromArray([["=METRIC()"]], {
    licenseKey: "gpl-v3"
  });

  const value = hfInstance.getCellValue({ sheet: 0, col: 0, row: 0 });

  console.log(value);

  return null;
}
```

**Comment by vhpoet** [comment]
ok that works, thanks for the detailed answer!

**Comment by sequba** [comment]
@vhpoet With **HyperFormula 2.0.1** we updated our guide about custom functions. You may want to have a look at it: https://handsontable.github.io/hyperformula/guide/custom-functions.html

**Comment by sequba** [comment]
@vhpoet, we've just released a new [custom functions guide](https://hyperformula.handsontable.com/guide/custom-functions.html) with an advanced example. You should check it out ;)
